## Imports

In [1]:
!pip install sacrebleu
!pip install evaluate meteor bert_score sacrebleu
!pip install -r requirements.txt

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 100.8/100.8 kB 13.5 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 113.0/113.0 kB 14.4 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.4/1.4 MB 90.6 MB/s eta 0:00:00
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 84.1/84.1 kB 10.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 62.7/62.7 MB 32.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 61.1/61.1 kB 8.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 12.3/12.3 MB 23.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 701.4/701.4 kB 59.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 65.5/65.5 kB 8.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 24.0/24.0 MB 25.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 44.1/44.1 kB 4.3 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.6/2.6 MB 109.0 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 57.0/57.0 kB 7.1 MB/s eta 0:00:00
Reason for being yanked: This version unfortunately does not work with 3.8 but we did not drop the support yet
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 10.0/10.0 MB 31.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.0/3.0 MB 118.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 77.1/77.1 kB 10.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 566.4/566.4 kB 57.3 MB/s eta 0:00:00
  Created wheel for sentencepiece: filename=sentencepiece-0.1.99-cp312-cp312-linux_x86_64.whl size=1266901 sha256=bbd3c6e2aea747268a7e278b42e82834c8bb3281e513d85d472cb7b1dee7c4b4
  Stored in directory: /root/.cache/pip/wheels/e0/8c/e0/65e33b1f4b8462dfc537a0cac02e5c03e1207564c300e4bde5
Succe

In [2]:
import requests
import re
import time
import os
import json
from pathlib import Path
import numpy as np
import torch
import torch.nn.functional as F
import sacrebleu
import evaluate
from tqdm import tqdm
from transformers import AutoTokenizer, AutoModel

import torch.nn as nn
from torch.utils.data import Dataset, DataLoader
from torch.nn.utils.rnn import pad_sequence

In [21]:
dirs = ["mined_code_summaries", "pt_files", "checkpoints"]

for d in dirs:
    Path(d).mkdir(parents=True, exist_ok=True)
    print(f"Verified: {os.path.abspath(d)}")

Verified: /content/mined_code_summaries
Verified: /content/pt_files
Verified: /content/checkpoints


In [25]:
# Insert your GitHub PAT here to avoid strict API limits
GITHUB_TOKEN = "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"
SIDE_CHECKPOINT = "required_files/side_model"

In [ ]:
HEADERS = {"Authorization": f"token {GITHUB_TOKEN}", "Accept": "application/vnd.github.v3+json"} if GITHUB_TOKEN != "YOUR_GITHUB_TOKEN" else {}

def fetch_top_java_repos(num_repos=50):
    """Fetches high-quality, production-grade Java repositories."""
    repos = []
    page = 1
    exclude_keywords = ["tutorial", "awesome", "interview", "leetcode", "learning", "demo", "practice"]

    print("Step 1a: Fetching Java repositories...")
    while len(repos) < num_repos:
        url = "https://api.github.com/search/repositories"
        query = "language:java stars:>1000 size:>10000 pushed:>2025-01-01 archived:false -topic:leetcode -topic:tutorial -topic:awesome"
        params = {"q": query, "sort": "stars", "order": "desc", "per_page": 100, "page": page}

        response = requests.get(url, headers=HEADERS, params=params)
        if response.status_code == 403:
            print("Rate limit hit during repo search. Sleeping 60s...")
            time.sleep(60)
            continue
        elif response.status_code != 200:
            break

        for item in response.json().get("items", []):
            if item.get("fork", False) or not item.get("description"):
                continue
            name, desc = item.get("name", "").lower(), item.get("description", "").lower()
            if any(kw in name or kw in desc for kw in exclude_keywords):
                continue

            repos.append(item["full_name"])
            if len(repos) >= num_repos:
                break
        page += 1
        time.sleep(2)
    return repos

def clean_summary(docstring):
    """Lowercases the summary and removes Javadoc formatting/newlines."""
    cleaned = re.sub(r'/\*\*|\*/|\*', '', docstring).lower().replace('\n', ' ').replace('\r', ' ')
    return re.sub(r'\s+', ' ', cleaned).strip()

def clean_code(code_string):
    """Flattens the Java method into a single line."""
    code_string = code_string.replace('\n', ' ').replace('\r', ' ')
    return re.sub(r'\s+', ' ', code_string).strip()

def extract_methods(java_code):
    """Finds Javadoc + method pairs and extracts the code body."""
    pairs = []
    pattern = re.compile(r'(/\*\*.*?\*/)\s*(?:@\w+(?:\([^)]*\))?\s*)*(?:public|private|protected)?\s*(?:static\s+)?(?:final\s+)?[\w\<\>\[\]\?]+\s+\w+\s*\([^)]*\)\s*(?:throws\s+[\w\s,]+)?\s*\{', re.DOTALL)

    for match in pattern.finditer(java_code):
        docstring = match.group(1)
        start_idx = match.end() - 1

        brace_count = 0
        end_idx = start_idx
        for i in range(start_idx, len(java_code)):
            if java_code[i] == '{': brace_count += 1
            elif java_code[i] == '}': brace_count -= 1
            if brace_count == 0:
                end_idx = i + 1
                break

        if brace_count == 0:
            method_code = java_code[match.start(0):end_idx]
            method_code = method_code.replace(docstring, '', 1).strip()
            summary = clean_summary(docstring)
            if len(summary.split()) > 3: # Keep only meaningful summaries
                pairs.append({"summary": summary, "code": clean_code(method_code)})
    return pairs

def build_dataset(target_train=50000, target_val=1000):
    repos = fetch_top_java_repos(num_repos=30) # 30 large repos should be enough for 51k methods
    all_pairs = []

    print(f"\nStep 1b: Mining Java files to collect {target_train + target_val} methods...")
    for repo in repos:
        url = f"https://api.github.com/repos/{repo}/git/trees/master?recursive=1"
        response = requests.get(url, headers=HEADERS)
        if response.status_code == 404:
            url = f"https://api.github.com/repos/{repo}/git/trees/main?recursive=1"
            response = requests.get(url, headers=HEADERS)

        tree = response.json().get('tree', [])
        java_urls = [f"https://raw.githubusercontent.com/{repo}/master/{item['path']}" for item in tree if item['path'].endswith('.java')]

        for i, raw_url in enumerate(java_urls):
            if len(all_pairs) >= (target_train + target_val): break
            try:
                if i % 50 == 0: time.sleep(1)
                raw_code = requests.get(raw_url).text
                all_pairs.extend(extract_methods(raw_code))

                if len(all_pairs) % 2000 == 0 and len(all_pairs) > 0:
                    print(f"Collected {len(all_pairs)} / {target_train + target_val} pairs...")
            except:
                continue

        if len(all_pairs) >= (target_train + target_val): break

    print("\nStep 1c: Splitting data and writing to .txt files...")
    output_dir = "mined_code_summaries"
    os.makedirs(output_dir, exist_ok=True)
    
    train_data = all_pairs[:target_train]
    val_data = all_pairs[target_train:target_train + target_val]

    def write_splits(data, code_file, summary_file):
        code_path = os.path.join(output_dir, code_file)
        summary_path = os.path.join(output_dir, summary_file)
        
        with open(code_path, 'w', encoding='utf-8') as fc, open(summary_path, 'w', encoding='utf-8') as fs:
            for p in data:
                fc.write(p['code'] + '\n')
                fs.write(p['summary'] + '\n')

    write_splits(train_data, 'train_code.txt', 'train_summary.txt')
    write_splits(val_data, 'val_code.txt', 'val_summary.txt')
    print("Phase 1 Complete! The 4 text files are ready for get_codet5_embeddings.py.")


build_dataset()

In [15]:
# Estimate token count by percentiles

def check_coverage(file_path):
    with open(file_path, 'r', encoding='utf-8') as f:
        # A rough estimate: 1 word usually equals ~1.2 to 1.5 tokens
        lengths = [len(line.split()) * 1.3 for line in f.readlines()]

    print(f"Stats for {file_path}:")
    print(f"  90% of sequences are under {np.percentile(lengths, 90):.0f} tokens")
    print(f"  95% of sequences are under {np.percentile(lengths, 95):.0f} tokens")
    print(f"  99% of sequences are under {np.percentile(lengths, 99):.0f} tokens")

check_coverage('mined_code_summaries/train_code.txt')
check_coverage('mined_code_summaries/train_summary.txt')
check_coverage('mined_code_summaries/val_code.txt')
check_coverage('mined_code_summaries/val_summary.txt')
check_coverage('required_files/sample_code.txt')
check_coverage('required_files/sample_summary.txt')

Stats for mined_code_summaries/train_code.txt:
  90% of sequences are under 112 tokens
  95% of sequences are under 176 tokens
  99% of sequences are under 373 tokens
Stats for mined_code_summaries/train_summary.txt:
  90% of sequences are under 162 tokens
  95% of sequences are under 286 tokens
  99% of sequences are under 840 tokens
Stats for mined_code_summaries/val_code.txt:
  90% of sequences are under 134 tokens
  95% of sequences are under 200 tokens
  99% of sequences are under 382 tokens
Stats for mined_code_summaries/val_summary.txt:
  90% of sequences are under 173 tokens
  95% of sequences are under 269 tokens
  99% of sequences are under 512 tokens
Stats for required_files/sample_code.txt:
  90% of sequences are under 55 tokens
  95% of sequences are under 56 tokens
  99% of sequences are under 57 tokens
Stats for required_files/sample_summary.txt:
  90% of sequences are under 18 tokens
  95% of sequences are under 19 tokens
  99% of sequences are under 20 tokens


In [16]:
# 1. Process the Training Data

# Tokenize and extract embeddings for the training code (max length 256)
!python required_files/get_codet5_embeddings.py --input mined_code_summaries/train_code.txt --output pt_files/train_code.pt --max_length 256

# Tokenize and extract embeddings for the training summaries (max length 512)
!python required_files/get_codet5_embeddings.py --input mined_code_summaries/train_summary.txt --output pt_files/train_summary.pt --max_length 512

# 2. Process the Validation Data
!python required_files/get_codet5_embeddings.py --input mined_code_summaries/val_code.txt --output pt_files/val_code.pt --max_length 256

!python required_files/get_codet5_embeddings.py --input mined_code_summaries/val_summary.txt --output pt_files/val_summary.pt --max_length 512

# 3. Process the Test Data
!python required_files/get_codet5_embeddings.py --input required_files/sample_code.txt --output pt_files/sample_code.pt --max_length 256

!python required_files/get_codet5_embeddings.py --input required_files/sample_summary.txt --output pt_files/sample_summary.pt --max_length 512

Loading tokenizer and model: Salesforce/codet5p-220m
2026-03-17 00:09:51.207395: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:467] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered
E0000 00:00:1773706191.229704    7316 cuda_dnn.cc:8579] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
E0000 00:00:1773706191.237082    7316 cuda_blas.cc:1407] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered
W0000 00:00:1773706191.255993    7316 computation_placer.cc:177] computation placer already registered. Please check linkage and avoid linking the same target more than once.
W0000 00:00:1773706191.256021    7316 computation_placer.cc:177] computation placer already registered. Please check linkage and avoid linking the same target more than once.
W0000 00:00:1773706191.256024    7

In [17]:
# Load the PyTorch dictionary files
train_code_data = torch.load("pt_files/train_code.pt")
train_summary_data = torch.load("pt_files/train_summary.pt")

val_code_data = torch.load("pt_files/val_code.pt")
val_summary_data = torch.load("pt_files/val_summary.pt")

# Extract the CodeT5+ Pretrained Embedding Matrix (32,100 x 768)
embedding_matrix = train_code_data['embedding_matrix']

token_ids = train_code_data['token_ids']
vocab_size = train_code_data['vocab_size']
embedding_dim = train_code_data['embedding_dim']

pad_id = train_code_data.get('pad_token_id', 0)
eos_id = train_code_data.get('eos_token_id', 2)
sos_id = train_code_data.get('bos_token_id', 1)

In [18]:
class SummarizationDataset(Dataset):
    def __init__(self, code_ids, summary_ids):
        self.code_ids = code_ids
        self.summary_ids = summary_ids

    def __len__(self):
        return len(self.code_ids)

    def __getitem__(self, idx):
        # Convert the lists of integers into PyTorch tensors
        return torch.tensor(self.code_ids[idx]), torch.tensor(self.summary_ids[idx])

def collate_fn(batch):
    code, summary = zip(*batch)
    # Pad sequences so they are the same length in the batch
    return (pad_sequence(code, batch_first=True, padding_value=pad_id),
            pad_sequence(summary, batch_first=True, padding_value=pad_id))

# Initialize DataLoaders
BATCH_SIZE = 32
HIDDEN_DIM = 256
MAX_LEN = 64

train_dataset = SummarizationDataset(train_code_data['token_ids'], train_summary_data['token_ids'])
val_dataset = SummarizationDataset(val_code_data['token_ids'], val_summary_data['token_ids'])

train_loader = DataLoader(train_dataset, batch_size=BATCH_SIZE, shuffle=True, collate_fn=collate_fn)
val_loader = DataLoader(val_dataset, batch_size=BATCH_SIZE, shuffle=False, collate_fn=collate_fn)

print(f"Train batches: {len(train_loader)} | Val batches: {len(val_loader)}")

Train batches: 1563 | Val batches: 32


In [19]:
class CodeSummarizationLSTM(nn.Module):
    def __init__(self, pretrained_matrix, hidden_dim, vocab_size):
        super().__init__()

        # LSTM with freeze=False
        self.embed = nn.Embedding.from_pretrained(pretrained_matrix, freeze=False)
        self.dropout = nn.Dropout(0.2)

        # Vector compresses into HIDDEN_DIM (256)
        embed_dim = pretrained_matrix.size(1)

        # Encoder and Decoder LSTMs
        self.encoder = nn.LSTM(embed_dim, hidden_dim, num_layers=2, batch_first=True, dropout=0.2)
        self.decoder = nn.LSTM(embed_dim, hidden_dim, num_layers=2, batch_first=True, dropout=0.2)

        # Output layer maps the 256-dim hidden state
        self.out = nn.Linear(hidden_dim, vocab_size)

    def forward(self, src, tgt):
        # src = code token IDs, tgt = summary token IDs
        src_emb = self.dropout(self.embed(src))
        tgt_emb = self.dropout(self.embed(tgt))

        _, hidden = self.encoder(src_emb)

        output, _ = self.decoder(tgt_emb, hidden)

        # Predict the next token logits
        return self.out(output)

    def generate(self, src):
        src_emb = self.dropout(self.embed(src))
        _, hidden = self.encoder(src_emb)

        outputs = []
        # Start generation with the SOS (Start of Sentence) token
        input_tok = torch.tensor([[sos_id]]).to(src.device)

        for _ in range(MAX_LEN):
            tgt_emb = self.dropout(self.embed(input_tok))
            out, hidden = self.decoder(tgt_emb, hidden)

            # Get the highest probability token ID
            pred = self.out(out).argmax(-1)

            # Stop if we predict EOS (End of Sentence)
            if pred.item() == eos_id:
                break

            outputs.append(pred.item())
            input_tok = pred

        return outputs

# Initialize the Model
DEVICE = torch.device('cuda' if torch.cuda.is_available() else 'cpu')

model = CodeSummarizationLSTM(
    pretrained_matrix=embedding_matrix,
    hidden_dim=HIDDEN_DIM,
    vocab_size=vocab_size
).to(DEVICE)

print(model)

CodeSummarizationLSTM(
  (embed): Embedding(32100, 768)
  (dropout): Dropout(p=0.2, inplace=False)
  (encoder): LSTM(768, 256, num_layers=2, batch_first=True, dropout=0.2)
  (decoder): LSTM(768, 256, num_layers=2, batch_first=True, dropout=0.2)
  (out): Linear(in_features=256, out_features=32100, bias=True)
)


In [20]:
# Initialize Optimizer and Loss
optimizer = torch.optim.Adam(model.parameters(), lr=0.001)
criterion = nn.CrossEntropyLoss(ignore_index=pad_id)

total_params = sum(p.numel() for p in model.parameters())
print(f"Model parameters: {total_params:,} total")

Model parameters: 36,056,420 total


In [22]:
tokenizer = AutoTokenizer.from_pretrained("Salesforce/codet5p-220m")

save_dir = "checkpoints"
save_name = "lstm_summarization_model"

# Setup Early Stopping Variables for BLEU-1
NUM_EPOCHS = 10
EVAL_STEPS = 500
best_bleu1 = 0.0
patience = 0
patience_limit = 5

steps_per_epoch = len(train_loader)
total_steps = NUM_EPOCHS * steps_per_epoch

step = 0
running_loss = 0
log_steps = 50

progress = tqdm(total=total_steps, desc="Training")

for epoch in range(NUM_EPOCHS):
    model.train()
    for src, tgt in train_loader:
        src, tgt = src.to(DEVICE), tgt.to(DEVICE)

        # Teacher forcing training
        output = model(src, tgt[:, :-1])
        loss = criterion(output.reshape(-1, output.size(-1)), tgt[:, 1:].reshape(-1))

        optimizer.zero_grad()
        loss.backward()
        torch.nn.utils.clip_grad_norm_(model.parameters(), 1.0)
        optimizer.step()

        step += 1
        running_loss += loss.item()

        if step % log_steps == 0:
            avg_loss = running_loss / log_steps
            progress.set_postfix({'train_loss': f'{avg_loss:.3f}', 'epoch': f'{step/steps_per_epoch:.2f}'})
            running_loss = 0

        progress.update(1)

        # Validation Phase (BLEU-1 Early Stopping)
        if step % EVAL_STEPS == 0:
            model.eval()
            predictions = []
            references = []

            with torch.no_grad():
                count = 0
                for val_src, val_tgt in val_loader:
                    val_src, val_tgt = val_src.to(DEVICE), val_tgt.to(DEVICE)

                    # Autoregressive generation evaluates one sequence at a time
                    for i in range(val_src.size(0)):
                        # Cap the evaluation subset to 200 sequences to prevent long evaluation times
                        # (Increase or remove this count limit if you want to evaluate the full 1000 val set)
                        if count >= 200:
                            break

                        pred_ids = model.generate(val_src[i:i+1])

                        # Decode the model's prediction
                        pred_text = tokenizer.decode(pred_ids, skip_special_tokens=True).strip()
                        if pred_text == "":
                            pred_text = "empty" # Prevent sacrebleu from crashing on empty predictions

                        # Decode the target reference (filtering out special tokens)
                        ref_ids = [idx.item() for idx in val_tgt[i] if idx.item() not in [pad_id, sos_id, eos_id]]
                        ref_text = tokenizer.decode(ref_ids, skip_special_tokens=True).strip()

                        predictions.append(pred_text)
                        references.append(ref_text)
                        count += 1

                    if count >= 200:
                        break

            # Calculate BLEU-1 using sacrebleu
            bleu_result = sacrebleu.corpus_bleu(predictions, [references])

            # The .precisions attribute holds [1-gram, 2-gram, 3-gram, 4-gram] precision.
            current_bleu1 = bleu_result.precisions[0]

            tqdm.write(f"Step {step}/{total_steps} | Validation BLEU-1: {current_bleu1:.2f}")

            # Early Stopping Logic: Save if BLEU-1 improves
            if current_bleu1 > best_bleu1:
                best_bleu1 = current_bleu1
                patience = 0
                torch.save(model.state_dict(), f'{save_dir}/{save_name}.pt')
                tqdm.write("  -> Model improved and saved!")
            else:
                patience += 1
                tqdm.write(f"  -> No improvement. Patience: {patience}/{patience_limit}")
                if patience >= patience_limit:
                    tqdm.write(f"\nEarly stopping triggered at step {step}!")
                    progress.close()
                    break # Break out of inner loop

            model.train() # Switch back to training mode

    if patience >= patience_limit:
        break # Break out of outer loop

progress.close()
print(f"\Best Validation BLEU-1: {best_bleu1:.2f}")

<>:110: SyntaxWarning: invalid escape sequence '\B'
<>:110: SyntaxWarning: invalid escape sequence '\B'
/tmp/ipykernel_1084/2666538552.py:110: SyntaxWarning: invalid escape sequence '\B'
  print(f"\Best Validation BLEU-1: {best_bleu1:.2f}")
/usr/local/lib/python3.12/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(
Training:   3%|▎         | 500/15630 [00:58<24:53, 10.13it/s, train_loss=5.982, epoch=0.32]

Step 500/15630 | Validation BLEU-1: 18.10


Training:   3%|▎         | 501/15630 [00:58<5:54:45,  1.41s/it, train_loss=5.982, epoch=0.32]

  -> Model improved and saved!


Training:   6%|▋         | 1001/15630 [01:56<10:05:09,  2.48s/it, train_loss=5.037, epoch=0.64]

Step 1000/15630 | Validation BLEU-1: 10.75
  -> No improvement. Patience: 1/5


Training:  10%|▉         | 1500/15630 [02:52<23:26, 10.05it/s, train_loss=4.757, epoch=0.96]

Step 1500/15630 | Validation BLEU-1: 21.06


Training:  10%|▉         | 1500/15630 [02:53<23:26, 10.05it/s, train_loss=4.757, epoch=0.96]

  -> Model improved and saved!


Training:  13%|█▎        | 2000/15630 [03:47<23:06,  9.83it/s, train_loss=4.433, epoch=1.28]

Step 2000/15630 | Validation BLEU-1: 25.09


Training:  13%|█▎        | 2001/15630 [03:48<5:27:06,  1.44s/it, train_loss=4.433, epoch=1.28]

  -> Model improved and saved!


Training:  16%|█▌        | 2500/15630 [04:42<22:29,  9.73it/s, train_loss=4.335, epoch=1.60]

Step 2500/15630 | Validation BLEU-1: 31.39


Training:  16%|█▌        | 2501/15630 [04:43<5:28:23,  1.50s/it, train_loss=4.335, epoch=1.60]

  -> Model improved and saved!


Training:  19%|█▉        | 3002/15630 [05:37<2:55:06,  1.20it/s, train_loss=4.191, epoch=1.92]

Step 3000/15630 | Validation BLEU-1: 30.89
  -> No improvement. Patience: 1/5


Training:  22%|██▏       | 3501/15630 [06:30<3:19:41,  1.01it/s, train_loss=4.008, epoch=2.24]

Step 3500/15630 | Validation BLEU-1: 30.90
  -> No improvement. Patience: 2/5


Training:  26%|██▌       | 4002/15630 [07:25<3:20:27,  1.03s/it, train_loss=3.930, epoch=2.56]

Step 4000/15630 | Validation BLEU-1: 28.44
  -> No improvement. Patience: 3/5


Training:  29%|██▉       | 4500/15630 [08:19<17:18, 10.72it/s, train_loss=3.858, epoch=2.88]

Step 4500/15630 | Validation BLEU-1: 41.30


Training:  29%|██▉       | 4501/15630 [08:19<3:00:53,  1.03it/s, train_loss=3.858, epoch=2.88]

  -> Model improved and saved!


Training:  32%|███▏      | 5001/15630 [09:14<3:22:59,  1.15s/it, train_loss=3.716, epoch=3.20]

Step 5000/15630 | Validation BLEU-1: 27.85
  -> No improvement. Patience: 1/5


Training:  35%|███▌      | 5500/15630 [10:08<13:30, 12.50it/s, train_loss=3.603, epoch=3.52]

Step 5500/15630 | Validation BLEU-1: 35.37
  -> No improvement. Patience: 2/5


Training:  38%|███▊      | 6000/15630 [11:03<16:06,  9.96it/s, train_loss=3.576, epoch=3.84]

Step 6000/15630 | Validation BLEU-1: 33.01
  -> No improvement. Patience: 3/5


Training:  42%|████▏     | 6501/15630 [11:58<2:38:34,  1.04s/it, train_loss=3.493, epoch=4.16]

Step 6500/15630 | Validation BLEU-1: 34.27
  -> No improvement. Patience: 4/5


Training:  45%|████▍     | 7000/15630 [12:52<15:52,  9.06it/s, train_loss=3.443, epoch=4.48]

Step 7000/15630 | Validation BLEU-1: 40.61
  -> No improvement. Patience: 5/5

Early stopping triggered at step 7000!
\Best Validation BLEU-1: 41.30


In [26]:
# Load the SIDE Metric Model
side_tokenizer = AutoTokenizer.from_pretrained(SIDE_CHECKPOINT)
side_model = AutoModel.from_pretrained(SIDE_CHECKPOINT).to(DEVICE)

def mean_pooling(model_output, attention_mask):
    """Mean Pooling - Takes attention mask into account for correct averaging"""
    token_embeddings = model_output[0]
    input_mask_expanded = attention_mask.unsqueeze(-1).expand(token_embeddings.size()).float()
    return torch.sum(token_embeddings * input_mask_expanded, 1) / torch.clamp(input_mask_expanded.sum(1), min=1e-9)

def compute_side_score(code_text, summary_text):
    """Computes the SIDE cosine similarity for a single code-summary pair"""
    pair = [code_text, summary_text]
    encoded_input = side_tokenizer(pair, padding=True, truncation=True, return_tensors='pt').to(DEVICE)

    with torch.no_grad():
        model_output = side_model(**encoded_input)

    sentence_embeddings = mean_pooling(model_output, encoded_input['attention_mask'])
    sentence_embeddings = F.normalize(sentence_embeddings, p=2, dim=1)

    # Calculate cosine similarity between the code [0] and summary [1]
    sim = torch.sum(sentence_embeddings[0] * sentence_embeddings[1]).item()
    return sim


# Load the Test Data and Checkpoint
test_code_data = torch.load("pt_files/sample_code.pt")
test_summary_data = torch.load("pt_files/sample_summary.pt")

test_dataset = SummarizationDataset(test_code_data['token_ids'], test_summary_data['token_ids'])
test_loader = DataLoader(test_dataset, batch_size=BATCH_SIZE, shuffle=False, collate_fn=collate_fn)

model.load_state_dict(torch.load(f'{save_dir}/{save_name}.pt', map_location=DEVICE))
model.eval()


# Generate Predictions and Compute Scores
predictions = []
references = []
side_scores = []

print("Generating predictions and calculating SIDE scores on the test set...")
with torch.no_grad():
    for src, tgt in tqdm(test_loader, desc="Testing"):
        src, tgt = src.to(DEVICE), tgt.to(DEVICE)

        for i in range(src.size(0)):
            # Generate token IDs
            pred_ids = model.generate(src[i:i+1])

            # Decode prediction
            pred_text = tokenizer.decode(pred_ids, skip_special_tokens=True).strip()
            if pred_text == "":
                pred_text = "empty"

            # Decode reference
            ref_ids = [idx.item() for idx in tgt[i] if idx.item() not in [pad_id, sos_id, eos_id]]
            ref_text = tokenizer.decode(ref_ids, skip_special_tokens=True).strip()

            # Decode the original source code (needed for SIDE)
            src_ids = [idx.item() for idx in src[i] if idx.item() not in [pad_id, sos_id, eos_id]]
            code_text = tokenizer.decode(src_ids, skip_special_tokens=True).strip()

            # Compute SIDE score for this prediction
            side_sim = compute_side_score(code_text, pred_text)
            side_scores.append(side_sim)

            predictions.append(pred_text)
            references.append(ref_text)

# Compute Final Aggregate Metrics ---
print("\n--- Final Evaluation Metrics ---")

# BLEU-1, 2, 3, 4
bleu = sacrebleu.corpus_bleu(predictions, [references])
print(f"BLEU-1: {bleu.precisions[0]:.2f}")
print(f"BLEU-2: {bleu.precisions[1]:.2f}")
print(f"BLEU-3: {bleu.precisions[2]:.2f}")
print(f"BLEU-4: {bleu.precisions[3]:.2f}")

# METEOR
meteor_metric = evaluate.load('meteor')
meteor_results = meteor_metric.compute(predictions=predictions, references=references)
print(f"METEOR: {meteor_results['meteor'] * 100:.2f}")

# BERTScore
bertscore_metric = evaluate.load('bertscore')
bs_results = bertscore_metric.compute(predictions=predictions, references=references, lang="en")
avg_bertscore = (sum(bs_results['f1']) / len(bs_results['f1'])) * 100
print(f"BERTScore (F1): {avg_bertscore:.2f}")

# SIDE
avg_side_score = (sum(side_scores) / len(side_scores)) * 100
print(f"SIDE: {avg_side_score:.2f}")

# Save and Display Sample Predictions ---
print("\n--- Sample Predictions ---")
for i in range(5):
    print(f"\n[Sample {i+1}]")
    print(f"Reference : {references[i]}")
    print(f"Generated : {predictions[i]}")

# Save all predictions to a JSON file for the repository
output_data = []
for i in range(len(predictions)):
    output_data.append({
        "id": i,
        "reference": references[i],
        "prediction": predictions[i]
    })

with open("test_predictions.json", "w") as f:
    json.dump(output_data, f, indent=4)

print("\nAll 1,000 predictions successfully saved to 'test_predictions.json'!")

Generating predictions and calculating SIDE scores on the test set...


Testing: 100%|██████████| 1/1 [00:00<00:00,  2.70it/s]



--- Final Evaluation Metrics ---
BLEU-1: 5.66
BLEU-2: 0.65
BLEU-3: 0.34
BLEU-4: 0.17


[nltk_data] Downloading package wordnet to /root/nltk_data...
[nltk_data] Downloading package punkt_tab to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt_tab.zip.
[nltk_data] Downloading package omw-1.4 to /root/nltk_data...


METEOR: 7.20


tokenizer_config.json:   0%|          | 0.00/25.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/482 [00:00<?, ?B/s]

vocab.json: 0.00B [00:00, ?B/s]

merges.txt: 0.00B [00:00, ?B/s]

tokenizer.json: 0.00B [00:00, ?B/s]

model.safetensors:   0%|          | 0.00/1.42G [00:00<?, ?B/s]

Some weights of RobertaModel were not initialized from the model checkpoint at roberta-large and are newly initialized: ['roberta.pooler.dense.bias', 'roberta.pooler.dense.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


BERTScore (F1): 83.54
SIDE: 34.37

--- Sample Predictions ---

[Sample 1]
Reference : concatenate two string arrays into one
Generated : a {@linkplain jittype} to be used to create a {@link string} with the given name. @param name the name of the property @param value the value to be used. @param value the value to be used. @return the value of the given value.

[Sample 2]
Reference : return true if the contents of the internal array bytes and the provided array data match
Generated : returns the value of the given value. @param value the value to be used. @param value the value to be used. @return the value

[Sample 3]
Reference : find the index of a search string within a string
Generated : returns the value of the given value. @param value the value to be used. @param value the value to be used. @return the value

[Sample 4]
Reference : close the socket connection
Generated : get the number of bytes in the given address. @return the number of bytes in the given address.

[Sample 5]


# Write Up

To start off, the pipeline began by mining public GitHub repositories via the GitHub API to construct a corpus of roughly 50,000 training and 1,000 validation code-summary pairs. The data was preprocessed by flattening the Java methods, filtering into a single whitespace-normalized line and lowercasing the summaries. From there, the provided get_codet5_embeddings.py script was used to tokenize the data and extract the pretrained CodeT5+ embedding matrix. The core architecture consisted of an Encoder-Decoder LSTM that was initialized with these code-aware embeddings so that no vocabulary building was needed. I utilized teacher forcing and an Adam optimizer during training, alongside an early stopping mechanism configured to halt training when the BLEU-1 score on the validation set stopped improving after many successions. Training ultimately triggered early stopping at step 7,000 (around epoch 4.5), achieving a peak validation BLEU-1 score of 41.30.

Despite good validation scores during training, the model's performance on the test set revealed significant limitations. The final evaluation yielded a BLEU-1 score of 5.66, a METEOR score of 7.20, and a SIDE score of 34.37. Interestingly, the BERTScore remained relatively high at 83.54, which is likely an artifact of overlapping contextual embeddings rather than a true reflection of semantic alignment. After looking at my model’s test predictions, it exposes some problems. The LSTM generated generic, boilerplate Javadoc phrases like "returns the value of the given value. @param value the value to be used" regardless of the input code's actual function. This stark drop from validation to testing indicates that the model heavily overfit to frequent patterns in the training data rather than genuinely learning to map code semantics to natural language. Moving forward, I would need to assess this issue by creating a more robust model or perhaps addressing issues in the preprocessing stages.